app/src/main/java/com/hedvig/app/datastore/SeenCrossSellsPreferencesDataMigration.kt

```diff
@@ -0,0 +1,45 @@
+package com.hedvig.app.datastore
+
+import androidx.datastore.core.DataMigration
+import androidx.datastore.preferences.core.Preferences
+import androidx.datastore.preferences.core.stringSetPreferencesKey
+import com.hedvig.android.owldroid.type.TypeOfContract
+import com.hedvig.app.service.badge.NotificationBadge
+
+class SeenCrossSellsPreferencesDataMigration : DataMigration<Preferences> {
+    private val oldCrossSellSetKey = stringSetPreferencesKey("SEEN_CROSS_SELLS")
+
+    override suspend fun shouldMigrate(currentData: Preferences): Boolean {
+        return try {
+            val oldCrossSellStrings = currentData[oldCrossSellSetKey] ?: return false
+            return oldCrossSellStrings.isNotEmpty()
+        } catch (e: ClassCastException) {
+            false
```

Comment:
Log error here

Reply:
The idea is that if the DataStore does contain data of the type Set<String> it means that it is storing information as we did before.

Optimally I wanted to check if it simply has data on the key `stringSetPreferencesKey("SEEN_CROSS_SELLS")` and that does return true (a function called .contains() on Preferences or smth like that), but unfortunately, this also returns true *after* the migration when the key is actually then `booleanPreferencesKey("SEEN_CROSS_SELLS")` so it just looks at the string but not the type contained. So with the .get approach, unfortunately it throws an exception and I am using it for the normal flow, not as an exceptional thing that would happen very rarely.

So this implementation seemed the only way available from the available API to see if the data stored in the datastore is of the old type, meaning with the key "SEEN_CROSS_SELLS" and at the same time containing the old set of strings.
So I'm using it just to run the migration when this is the case.